datafusion/core/src/physical_optimizer/remove_unnecessary_sorts.rs

```diff
@@ -0,0 +1,887 @@
+//! physical plan and removes the ones it can prove unnecessary. The rule can
+//! work on valid *and* invalid physical plans with respect to sorting
+//! requirements, but always produces a valid physical plan in this sense.
+//!
+//! A non-realistic but easy to follow example: Assume that we somehow get the fragment
+//! "SortExec: [nullable_col@0 ASC]",
+//! "  SortExec: [non_nullable_col@1 ASC]",
+//! in the physical plan. The first sort is unnecessary since its result is overwritten
+//! by another SortExec. Therefore, this rule removes it from the physical plan.
+use crate::error::Result;
+use crate::physical_optimizer::utils::{
+    add_sort_above_child, ordering_satisfy, ordering_satisfy_concrete,
+};
+use crate::physical_optimizer::PhysicalOptimizerRule;
+use crate::physical_plan::rewrite::TreeNodeRewritable;
+use crate::physical_plan::sorts::sort::SortExec;
+use crate::physical_plan::windows::WindowAggExec;
+use crate::physical_plan::{with_new_children_if_necessary, ExecutionPlan};
+use crate::prelude::SessionConfig;
+use arrow::datatypes::SchemaRef;
+use datafusion_common::{reverse_sort_options, DataFusionError};
+use datafusion_physical_expr::{PhysicalExpr, PhysicalSortExpr};
+use itertools::izip;
+use std::iter::zip;
+use std::sync::Arc;
+
+/// This rule inspects SortExec's in the given physical plan and removes the
+/// ones it can prove unnecessary.
+#[derive(Default)]
+pub struct RemoveUnnecessarySorts {}
+
+impl RemoveUnnecessarySorts {
+    #[allow(missing_docs)]
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+/// This is a "data class" we use within the [RemoveUnnecessarySorts] rule
+/// that tracks the closest `SortExec` descendant for every child of a plan.
+#[derive(Debug, Clone)]
+struct PlanWithCorrespondingSort {
+    plan: Arc<dyn ExecutionPlan>,
+    // For every child, keep a vector of `ExecutionPlan`s starting from the
+    // closest `SortExec` till the current plan. The first index of the tuple is
+    // the child index of the plan -- we need this information as we make updates.
+    sort_onwards: Vec<Vec<(usize, Arc<dyn ExecutionPlan>)>>,
+}
+
+impl PlanWithCorrespondingSort {
+    pub fn new(plan: Arc<dyn ExecutionPlan>) -> Self {
+        let length = plan.children().len();
+        PlanWithCorrespondingSort {
+            plan,
+            sort_onwards: vec![vec![]; length],
+        }
+    }
+
+    pub fn children(&self) -> Vec<PlanWithCorrespondingSort> {
+        self.plan
+            .children()
+            .into_iter()
+            .map(|child| PlanWithCorrespondingSort::new(child))
+            .collect()
+    }
+}
+
+impl TreeNodeRewritable for PlanWithCorrespondingSort {
+    fn map_children<F>(self, transform: F) -> Result<Self>
+    where
+        F: FnMut(Self) -> Result<Self>,
+    {
+        let children = self.children();
+        if children.is_empty() {
+            Ok(self)
+        } else {
+            let children_requirements = children
```

Comment:
I was expecting to see a check like the following in this module



```rust
// if there is a sort node child
if node.child.is_sort() {
  let sort = node.child;
  // if the sort's input ordering already satisfies the output
  // required ordering, it can be removed
  if ordering_satisfy_concrete(
    sort.child.output_ordering()
    node.required_input_ordering()
  ) {
    // remove sort and connect child directly to node
    node.child = sort.child
}
```

Reply:
Indeed we are doing this check. However, we are removing `SortExec`s from multi layer links, such as `Sort` -> `OrderMaintainingExec`-> `AnotherOrderMaintainingExec` ->`OrderRequiringExec` corresponding snippet is 
[here](https://github.com/synnada-ai/arrow-datafusion/blob/572a1a4271651ec0f5223389c9d1ab694e476c1f/datafusion/core/src/physical_optimizer/remove_unnecessary_sorts.rs#L173-L183)

Reply:
In that case I would expect (hope?) that the top `OrderMaintaingExec` had the same `output_ordering` as the `OrderRequiringExec` 

I wonder if for some types of OrderMaintaingExec nodes the "output_ordering" isn't properly computed / communicated 🤔

Reply:
> I wonder if for some types of OrderMaintaingExec nodes the "output_ordering" isn't properly computed / communicated 🤔

This is right some Executors such as `ProjectionExec`, `WindowAggExec` modifies `PhysicalExpr`s (when it is `Column`). Directly comparing `required_input_ordering` and `output_ordering` may not produce what is desired. We plan to file another PR to fix this. We added a [TODO](https://github.com/synnada-ai/arrow-datafusion/blob/572a1a4271651ec0f5223389c9d1ab694e476c1f/datafusion/core/src/physical_optimizer/remove_unnecessary_sorts.rs#L197-L200) mentioning this observation.